netket/logging/hdf5_log.py

```diff
@@ -0,0 +1,184 @@
+import os
+import numpy as np
+from ..logging import RuntimeLog
+from flax.serialization import to_bytes
+
+try:
+    import h5py
+except ImportError:
+    pass
+
+
+_mode_shorthands = {"write": "w", "append": "a", "fail": "x"}
+
+
+def tree_log(tree, root, data, *, iter=None):
+    """
+    Maps all elements in tree, recursively calling tree_log with a new root string,
+    and when it reaches leaves adds them to the `data` inplace.
+
+    Args:
+        tree: a pytree where the leaf nodes contain data
+        root: the root of the tags used to log to HDF5
+        data: an HDF5 file modified in place
+        iter: an interger number specifying at which iteration the data was generated
+    """
+
+    if tree is None:
+        return
+
+    elif isinstance(tree, list):
+        for (i, val) in enumerate(tree):
+            tree_log(val, f"{root}/{i}", data, iter=iter)
+
+    # handle namedtuples
+    elif isinstance(tree, list) and hasattr(tree, "_fields"):
+        tree_log(iter, f"{root}/iter", data)
+        for key in tree._fields:
+            tree_log(getattr(tree, key), f"{root}/{key}", data)
+
+    elif isinstance(tree, tuple):
+        tree_log(iter, f"{root}/iter", data)
+        for (i, val) in enumerate(tree):
+            tree_log(val, f"{root}/{i}", data)
+
+    elif isinstance(tree, dict):
+        for key, value in tree.items():
+            tree_log(value, f"{root}/{key}", data, iter=iter)  # noqa: F722
+
+    elif hasattr(tree, "to_compound"):
+        tree_log(iter, f"{root}/iter", data)
+        tree_log(tree.to_compound()[1], root, data)  # noqa: F722
+
+    elif hasattr(tree, "to_dict"):
+        tree_log(iter, f"{root}/iter", data)
+        tree_log(tree.to_dict(), root, data)  # noqa: F722
+
+    else:
+        if iter is not None:
+            tree_log(iter, f"{root}/iter", data)
+            root = f"{root}/value"
+        value = np.asarray(tree)
+        if root in data:
+            f_value = data[root]
+            f_value.resize(f_value.shape[0] + 1, axis=0)
+            f_value[-1] = value
+        else:
+            maxshape = (None,) + value.shape
+            data.create_dataset(root, data=[value], maxshape=maxshape)
+
+
+class HDF5Log(RuntimeLog):
```

Comment:
Should I remove the inheritance on the runtime logger?

Reply:
while you're at it can you also run black on the branch? I think some of my commits broke the formatting.

Reply:
Yup, will do